### PR TITLE
Added feature for multiple oidc federation

### PR DESCRIPTION
**Pull request by narengan**
Added functionality for federating with multiple OIDC Providers.

### CODE REVIEW

File: roles/keystone/tasks/openidc.yml [hunk 1]
```diff
@@ -26,3 +26,45 @@
...
+    src: etc/apache2/openidc/metadata/template.provider
+    dest: /etc/apache2/openidc/metadata/{{ item.issuer | regex_replace('https://','') }}.provider
+    owner: www-data
+    group: root
+    mode: 0640
+  when: item.provider_metadata_url is undefined
+  with_items: "{{ keystone.federation.sp.oidc.providers_info }}"
+
+- name: create .provider file for the IdPs
+  shell: "curl {{ item.provider_metadata_url }}  > /etc/apache2/openidc/metadata/{{ item.issuer | regex_replace('https://','') }}.provider"
```
Comment (narengan) [inline roles/keystone/tasks/openidc.yml]: TODO: handle curl command failure

File: roles/keystone/templates/etc/apache2/sites-available/keystone.conf [hunk 2]
```diff
@@ -16,37 +16,49 @@ Listen {{ endpoints.keystone.port.backend_api }}
...
-    OIDCProviderAuthorizationEndpoint {{ keystone.federation.sp.oidc.authorization_endpoint }}
-    OIDCProviderTokenEndpoint {{ keystone.federation.sp.oidc.token_endpoint }}
-    OIDCClientID {{ keystone.federation.sp.oidc.client_id }}
-    OIDCClientSecret {{ keystone.federation.sp.oidc.client_secret }}
     OIDCCryptoPassphrase {{ keystone.federation.sp.oidc.crypto_passphrase }}
     OIDCRedirectURI {{ keystone.federation.sp.oidc.redirect_uri }}
 
     # oauth
-    {% if not keystone.federation.sp.oidc.ssl_validate_server|bool -%}
+
```
Comment (lihkin213) [inline roles/keystone/templates/etc/apache2/sites-available/keystone.conf]: Would there be a way to handle additional parameters here. There could be other oauth directives that may be needed for specific providers.

File: roles/haproxy/templates/etc/haproxy/haproxy_openstack.cfg [hunk 3]
```diff
@@ -149,6 +149,9 @@ frontend {{ name }}
   bind :::{{ enc_port }} ssl crt /etc/haproxy/openstack.pem no-sslv3 ciphers AES128-SHA:AES256-SHA
   default_backend {{ name }}
   reqadd X-Forwarded-Proto:\ https
+  {% if name=="keystone" -%}
+  reqadd X-Forwarded-Port:\ 5000
```
Comment (edtubillara) [inline roles/haproxy/templates/etc/haproxy/haproxy_openstack.cfg]: Why was this added again? Did this work before with Blue Box panel?

File: roles/keystone/tasks/openidc.yml [hunk 4]
```diff
@@ -26,3 +26,45 @@
...
   apache2_module: name=auth_openidc
   notify: reload apache
+
+- name: create metadata folder for multiple oidc configuration files
+  file:
+    path: "/etc/apache2/openidc/metadata/"
+    state: directory
+    owner: www-data
+    group: root
+    mode: 0775
```
Comment (edtubillara) [inline roles/keystone/tasks/openidc.yml]: Can we make this 0770? Is there a process/user that needs to read this that isn't Apache (www-data)?

File: roles/keystone/tasks/openidc.yml [hunk 5]
```diff
@@ -26,3 +26,45 @@
...
+    recurse: yes
+
+- name: create .client file for the IdPs
+  template:
+    src: etc/apache2/openidc/metadata/template.client
+    dest: /etc/apache2/openidc/metadata/{{ item.issuer | regex_replace('https://','') }}.client
```
Comment (edtubillara) [inline roles/keystone/tasks/openidc.yml]: doing the regex_replace seems a little weird... Can we just have a variable for this which has characters already safe for folder names?

File: roles/keystone/tasks/openidc.yml (hunk 1 above)
Comment (edtubillara) [inline roles/keystone/tasks/openidc.yml]: Does every IDP have a metadata_url? What did we do for Blue Box Panel?

**Comment by edtubillara** [comment]
nice job. please see my comments

File: roles/haproxy/templates/etc/haproxy/haproxy_openstack.cfg (hunk 3 above)
Comment (lihkin213) [inline roles/haproxy/templates/etc/haproxy/haproxy_openstack.cfg]: This is needed because of the way we are doing multiple OP providers. The configuration for openid-connect is stored in files and because of this the client is redirected to port 5002 instead of the external port 5000, after the OIDC handshake done is.

File: roles/keystone/tasks/openidc.yml (hunk 1 above)
Comment (narengan) [inline roles/keystone/tasks/openidc.yml]: Mostly all IDP have an metadata_url. Since Bluebox Panel doesnt have one I check if the metadata variable is set and there are two flows from there

File: roles/keystone/tasks/openidc.yml (hunk 4 above)
Comment (narengan) [inline roles/keystone/tasks/openidc.yml]: there is no daemon. According to documentation only www-data accesses the directory. I can change the permissions

File: roles/keystone/tasks/openidc.yml (hunk 1 above)
Comment (edtubillara) [inline roles/keystone/tasks/openidc.yml]: This is an alternative to curl: http://docs.ansible.com/ansible/get_url_module.html

Comment (narengan) [inline roles/keystone/tasks/openidc.yml]: Tried that and got an error: failed to create temporary content file: (-1, 'Unexpected EOF')"

File: roles/keystone/templates/etc/apache2/sites-available/keystone.conf (hunk 2 above)
Comment (narengan) [inline roles/keystone/templates/etc/apache2/sites-available/keystone.conf]: I have added the optional parameters as well

File: roles/keystone/tasks/openidc.yml [hunk 6]
```diff
@@ -26,3 +26,45 @@
...
+
+- name: create .provider file for the IdPs
+  template:
+    src: etc/apache2/openidc/metadata/template.provider
+    dest: /etc/apache2/openidc/metadata/{{ item.metadata_file_name }}.provider
+    owner: www-data
+    group: root
+    mode: 0640
+  when: item.provider_metadata_url is undefined
+  with_items: "{{ keystone.federation.sp.oidc.providers_info }}"
```
Comment (j2sol) [inline roles/keystone/tasks/openidc.yml]: You could collapse the three tasks above into one task using a with_together or some other way to iterate over two sets, to iterate over the providers_info and client|conf|provider files. Just an optimization, but not important.

File: roles/keystone/tasks/openidc.yml [hunk 7]
```diff
@@ -26,3 +26,45 @@
...
+
+- name: create .provider file for the IdPs
+  shell: "curl --fail {{ item.provider_metadata_url }}  > /etc/apache2/openidc/metadata/{{ item.metadata_file_name }}.provider"
```
Comment (j2sol) [inline roles/keystone/tasks/openidc.yml]: It'd be interesting to dig into why the `url` or `uri` module didn't work here.

**Comment by j2sol** [comment]
Looks good enough for me. I'll merge it, we can modify it as needed.
